docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If not defined,
+  // this NetworkPolicyPort does not restrict based on port number.
+  // If defined, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicySource struct {
+  // If 'Namespaces' is defined, 'Pods' must not be.
+  // This is a label selector which selects Pods in this namespace.
+  // This NetworkPolicySource matches any pods selected by this selector.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // If 'Pods' is defined, 'Namespces' must not be.
+  // Selects Kubernetes Namespaces.  This NetworkPolicySource matches 
+  // all pods in all namespaces selected by this label selector. 
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+### Behavior
+
+The following pseudo-code attempts to define when traffic is allowed to a given pod when using this API.
+
+```python
+def is_traffic_allowed(traffic, pod):
+  """
+  Returns True if traffic is allowed to this pod, False otherwise.
+  """
+  if namespace.network_isolation == False:
+    # If network-isolation is disabled, all traffic is allowed.
+    return True 
+  elif traffic.source == pod.node:
+    # Traffic is from the pod's host - this allows for kubelet health checks.
+    return True
+  else:
+    # If network-isolation is enabled, only allow traffic 
+    # that matches a network policy which selects this pod.
+    for network_policy in all_network_policies:
+      if not network_policy.podSelector.selects(pod):
+        # This policy doesn't select this pod. Try the next one. 
+        continue
+
+      # This policy selects this pod.  Check each ingress rule 
+      # defined on this policy to see if it allows the traffic.
+      for ingress_rule in network_policy.ingress:
+        if ingress_rule.from.matches(traffic) and \
+             ingress_rule.ports.matches(traffic):
+          return True 
+
+  # network-isolation is True and no policies match the given pod and traffic.
+  return False
+```
+
+### Open Questions
+
+> The question then remains whether one NP object per "app" makes sense (in this case app={a,b,c}) or whether one NP object per selector
+
+- A single podSelector per NetworkPolicy may lead to managing a large number of NetworkPolicy objects, each of which is small and easy to understand on its own.  Allowing a podSelecor per ingress rule would allow for fewer, larger NetworkPolicy objects that may be easier to manage, but harder to understand individually.  This proposal has opted to favor a larger number of smaller objects that are easier to understand.  Is this the right decision? 
+
+- Is the `Namespaces` selector in the `NetworkPolicySource` struct too coarse? Do we need to support the AND combination of `Namespaces` and `Pods`? 
```

Comment:
@MikeSpreitzer - Sorry about that, totally misread your point!

So the options here would be:
* For a given `NetworkPolicySource`, no more than and no fewer than ONE of [`Pods`, `Namespaces`] can be defined.
* To indicate "from all sources", you leave the `From` list empty.
e.g 
```yaml
ingress:
  - ports:
      - port: 80
```
> Note that while "from" is missing in the YAML, the API defaults that field to an empty list.

OR

* For a given `NetworkPolicySource` no more than ONE of [`Pods`, `Namespaces`] can be defined.
* To indicate "from all sources", you define a single `NetworkPolicySource` with neither `Pods` nor `Namespaces` defined.
e.g
```yaml
ingress:
  - from:
      - {}
    ports:
      - port: 80
```

The latter option is more verbose to express in yaml, but I think I agree that for consistency, an empty `From` list should allow nothing, with each element in the list then adding allowed sources.

This is a departure from the original ThirdPartyResource schema we agreed, so I'm going to wait until we can get @thockin's sign off on this before updating the doc.